overlord/fdestate/extop.go

```diff
@@ -0,0 +1,162 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package fdestate
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// ExternalOperationStatus captures the status of operations running externally.
+type ExternalOperationStatus int
+
+// Admitted status values for changes and tasks.
+const (
+	DefaultStatus ExternalOperationStatus = 0
+
+	// Preparing means that we are performing preparation steps, but the
+	// operation isn't yet running externally.
+	PreparingStatus ExternalOperationStatus = 1
+
+	// DoingStatus means the operation is running externally.
+	DoingStatus ExternalOperationStatus = 2
+
+	// DoneStatus means the operation has completed successfully.
+	DoneStatus ExternalOperationStatus = 3
+
+	// CompletingStatus means the operation is completing.
+	CompletingStatus ExternalOperationStatus = 4
+
+	// AbortingStatus means the operation is aborting.
+	AbortingStatus ExternalOperationStatus = 5
+
+	// ErrorStatus means the operation has failed.
+	ErrorStatus ExternalOperationStatus = 6
+)
+
+func (s ExternalOperationStatus) String() string {
+	switch s {
+	case DefaultStatus:
+		return "default"
+	case PreparingStatus:
+		return "preparing"
+	case DoingStatus:
+		return "doing"
+	case DoneStatus:
+		return "done"
+	case CompletingStatus:
+		return "completing"
+	case AbortingStatus:
+		return "aborting"
+	case ErrorStatus:
+		return "error"
+	}
+	panic(fmt.Sprintf("internal error: unexpected external operation status code: %d", s))
+}
+
+type externalOperation struct {
+	Kind string `json:"kind"`
+	// ChangeID is ID of a state.Change associated with the operation.
+	ChangeID string                  `json:"change-id"`
+	Status   ExternalOperationStatus `json:"status"`
+	// Err when not empty, carries the error message, usually associated with
+	// the operation having an ErrorStatus or AbortingStatus.
+	Err string `json:"err"`
+	// Context is an opaque piece of data associated with the operation.
+	Context json.RawMessage `json:"context"`
+}
+
+func (e *externalOperation) SetFailed(msg string) {
+	e.Status = ErrorStatus
+	e.Err = msg
+}
+
+func (e *externalOperation) SetStatus(status ExternalOperationStatus) {
+	e.Status = status
+}
+
+// Equal checks the operations for equality by comparing their kind and the
+// parent change ID.
+func (e *externalOperation) Equal(other *externalOperation) bool {
+	return e.ChangeID == other.ChangeID && e.Kind == other.Kind
+}
+
+func (e *externalOperation) IsReady() bool {
+	switch e.Status {
+	case DoneStatus, ErrorStatus:
+		return true
+	}
+
+	return false
+}
+
+func findPendingExternalOperationByKind(st *state.State, kind string) (foundOp *externalOperation, err error) {
```

Comment:
we probably need a comment somewhere that we expect at most one non-ready op per kind?

Reply:
or should try to find the newest instead?

Reply:
looking at the test it seems you want to support having many but finding the first? if that's really the case this need to be explained in a comment to the function

Reply:
I'll leave a comment and will tweak the name. The intent is indeed to find the first pending, which in fact there should only be one of. In any case, with the API as it is, it is possible to add more than one, so the point is valid.